### PR TITLE
docs/design: add proposal for recursive common table expression (CTE)

### LINKED ISSUE #17472: support common table expression (CTE)
## Description

**Is your feature request related to a problem? Please describe:**

Related issues:
- https://github.com/pingcap/tidb/issues/6824 request to support recursive common table expression
- https://github.com/pingcap/tidb/issues/17916 request to support common table expression to run TPC-DS and other analytic workloads
- https://github.com/pingcap/parser/issues/289 request to support common table expression to be compatible with MySQL 8.0

**Describe the feature you'd like:**

Support the common table expression feature like MySQL [13.2.15 WITH (Common Table Expressions)](https://dev.mysql.com/doc/refman/8.0/en/with.html):

> A common table expression (CTE) is a named temporary result set that exists within the scope of a single statement and that can be referred to later within that statement, possibly multiple times. The following discussion describes how to write statements that use CTEs.

There are two kinds of CTEs:

- non-recursive CTE
- recursive CTE

A typical query example of non-recursive CTE:

```sql
WITH
  cte1 AS (SELECT a, b FROM table1),
  cte2 AS (SELECT c, d FROM table2)
SELECT b, d FROM cte1 JOIN cte2
WHERE cte1.a = cte2.c;
```

A typical query example of recursive CTE:

```sql
WITH RECURSIVE cte (n) AS
(
  SELECT 1
  UNION ALL
  SELECT n + 1 FROM cte WHERE n < 5
)
SELECT * FROM cte;
```

### PR DESCRIPTION
**Pull request by pingyu**
### What problem does this PR solve?

Issue Number:  #17472 #6824 

Problem Summary:

### What is changed and how it works?

Proposal: [Recursive Common Table Expression](https://github.com/pingyu/tidb/blob/executor_recursive_cte/docs/design/2020-08-08-recursive-common-table-expression.md) 

What's Changed:
(TBD)

How it Works:
(TBD)

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`: (TBD)

### Check List 

Tests 
(TBD)
- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects
(TBD)
- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

- Support recursive common table expression (CTE)

### CODE REVIEW

File: docs/design/2020-08-08-recursive-common-table-expression.md [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+
+## Abstract
+
+This proposal proposes to support the __Recursive Common Table Expression (Recursive CTE)__.
+
+
+## Background
+
+From MySQL User Document, [13.2.15 WITH (Common Table Expressions)](https://dev.mysql.com/doc/refman/8.0/en/with.html):
```
Comment (zz-jason) [inline docs/design/2020-08-08-recursive-common-table-expression.md:16]: It's better to describe more about the usage of recursive CTE like what we did in the proposal of non-recursive CTE.

File: docs/design/2020-08-08-recursive-common-table-expression.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+At the same time, MariaDB[4] utilize two temporary tables to save results:
+* Records for final result.
+* Records for latest iteration.
+
+
+## Implementation
+
+### Temporary Table
+Temporary table is implemented by `chunk.RowContainer`, which stores tuples in memory, or spill to disk if exceeds memory quota.
```
Comment (zz-jason) [inline docs/design/2020-08-08-recursive-common-table-expression.md:81]: Is it possible to share the same code which implements the temporary table described in https://github.com/pingcap/tidb/issues/1248?

File: docs/design/2020-08-08-recursive-common-table-expression.md [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+|    3 |
+|    4 |
+|    5 |
++------+
+```
+
+## Proposal
+
+### Seed part & Recursive part
+Recursive CTE has two parts, seed part and recursive part, separated by `UNION ALL` or `UNION [DISTINCT]`.
```
Comment (zz-jason) [inline docs/design/2020-08-08-recursive-common-table-expression.md:46]: It's better to introduce the syntax of recursive CTE beforing explaining what are `seed`, `union all`, and `recursive` parts.

**Comment by zz-jason** [comment]
I'm going to close this PR since it hasn't been updated for a long time. feel free to reopen if you want to continue with it. thank you for your contribution.

**Comment by guo-shaoge** [comment]
The design doc is great, and we need to continue to finish the implementation of recursive CTE.
I'd like to open a new PR and add some details based on your design doc. Is that OK? @pingyu

**Comment by pingyu** [comment]
> The design doc is great, and we need to continue to finish the implementation of recursive CTE.
> I'd like to open a new PR and add some details based on your design doc. Is that OK? @pingyu

That's great !

**Pull request closed without merging**